fix(airdrop): fix decimals validation

friofry opened the pull request:
Fixes #11918

### What does the PR do

1) HoldingsDropdown.qml - add "decimal" to assetPanel ListModel
2) InlineNetworksComboBox.qml - propogate "decimal" from the model  
3) TokenPanel.qml - propogate "decimal" from to AmountInput
4) AmountInput.qml - validate number of digits in franctional part of entered amount
### Affected areas

Community/Airdrop

### StatusQ checklist

- [ ] add documentation if necessary (new component, new feature)
- [ ] update sandbox app
  - in case of new component, add new component page
  - in case of new features, add variation to existing component page
  - nice to have: add it to the demo application as well
- [x] test changes in both light and dark theme?

### Screenshot of functionality (including design for comparison)
<kbd>
<img width="317" alt="image" src="https://github.com/status-im/status-desktop/assets/1083341/39eecb5e-859f-458d-8704-6fcfc6aaf368"></kbd>

- [ ] I've checked the design and this PR matches it

noeliaSD commented on ui/imports/shared/controls/AmountInput.qml:
```diff
@@ -10,14 +10,15 @@ import utils 1.0
     id: root
 
-    property int maximumLength: 10
+    property int maximumLength: 10 + tokenDecimals
```
I'm not sure about that.  I think the maximum allowed digits should be 10 including the decimals. Probably a question for @benjthayer ?

friofry replied:
Hi @noeliaSD 

I added numbers of decimals to HoldingsDropdownPage.qml and AirdropsSettingsPanelPage.qml for testing purposes

<img width="799" alt="image" src="https://github.com/status-im/status-desktop/assets/1083341/31052570-cf37-44a4-8866-5caa29a02ac8">

<img width="803" alt="image" src="https://github.com/status-im/status-desktop/assets/1083341/17b8ae68-90af-48d0-8e15-bf4963cfb545">

(I skipped PermissionsSettingsPanelPage.qml because if I follow the same approach, the UI will change a bit, so I need to investigate this further. If you don't mind, I'd suggest doing this separately)

noeliaSD replied:
Thanks for adding it to `storybook`!

noeliaSD replied:
> I'm not sure about that. I think the maximum allowed digits should be 10 including the decimals. Probably a question for @benjthayer ?

Regarding this comment, I've been talking with `Ben` about this digits restriction and after checking the requirements his last design/requirements are the following:

1) For community assets we can validate if the user enters a larger number of `decimal points` than the specified number of `decimal points` for that asset (specified during the community asset minting flow)
2) No other restriction on digits or decimals other than what fits in the input. 

There are 2 specific states added in the [designs](https://www.figma.com/file/qHfFm7C9LwtXpfdbxssCK3/Kuba%E2%8E%9CDesktop---Communities?type=design&node-id=52547-86075&mode=design&t=pdq48KFhbmfDZiQz-4) to cover them:

![image](https://github.com/status-im/status-desktop/assets/97019400/3910bd28-3870-4da7-9429-5e37b0b85baa)

**Summary:**
- We need to update the error message text when validating decimals.
- We don't need to add validation of 10 digits but blocking the user to input more digits than the ones that fit in the input (AND no error displayed when that happens).

Hope that's clear! :blush: 

cc: @benjthayer

friofry replied:
Okay, thanks for the clarification! I'll fix it

friofry replied:
I've updated the PR 

The udpated copy:
![Screenshot 2024-03-13 at 18 10 19](https://github.com/status-im/status-desktop/assets/1083341/bd919830-d388-4783-b611-5021155821bb)

as per discussed with @benjthayer possible solution for the Input field without the character limit:
https://github.com/status-im/status-desktop/assets/1083341/bcbf062b-6a22-45df-9321-06911a0af927
^
(the number of characters may depend on the DPI of the screen, or on custom fonts. For example, at 100% will fit 15 digits, but at 250% only 7. I would suggest not limiting the number of digits in width, because it may work differently on different computers)

noeliaSD replied:
> Hi @noeliaSD please check the updated PR
> 
> > We need to update the error message text when validating decimals
> 
> The updated copy (additionally I've aligned the validation error to the left):
> 

Thanks for your update! I can see this warning on `storybook` that would worth it if you can take a look:

https://github.com/status-im/status-desktop/assets/97019400/65919c78-4559-441d-83a4-ee6a27389482

> (the number of characters may depend on the DPI of the screen, or on custom fonts. For example, at 100% will fit 15 digits, but at 250% only 7. I would suggest not limiting the number of digits in width, because it may work differently on different computers)

I've asked again to `Ben` to see how he sees the solution proposed! Thanks you!

noeliaSD replied:
>  Screen.Recording.2024-03-19.at.12.20.51.mov 
> > (the number of characters may depend on the DPI of the screen, or on custom fonts. For example, at 100% will fit 15 digits, but at 250% only 7. I would suggest not limiting the number of digits in width, because it may work differently on different computers)
> 
> I've asked again to `Ben` to see how he sees the solution proposed! Thanks you!

Ok, so, _Ben_ is ok with this solution right now, but we will create a separate task to add lateral scrolling into status input so it's easier for the user to check the amount not only using the arrows: https://github.com/status-im/status-desktop/issues/14032

friofry replied:
the warning is fixed

noeliaSD reviewed (commented):
In general LGTM! I've left a comment bc I'm not sure about something and mentioned the designer!

And it would be nice if you update the `storybook` assets related models with a decimal value so we can test it also there. For example in these pages (probably these 3 examples share the assets models, so just 1 update):
<img width="1233" alt="Screenshot 2024-03-08 at 12 34 13" src="https://github.com/status-im/status-desktop/assets/97019400/ae88f6b6-c68c-4f9f-8a2d-99dbb8cc6226">
<img width="1233" alt="Screenshot 2024-03-08 at 12 33 59" src="https://github.com/status-im/status-desktop/assets/97019400/150b6800-753f-42e5-9f55-7b43ec05a01f">
<img width="1233" alt="Screenshot 2024-03-08 at 12 33 14" src="https://github.com/status-im/status-desktop/assets/97019400/d42b7e99-b9f7-485a-a5dc-a40c553b0cc9">

noeliaSD commented on ui/imports/shared/controls/AmountInput.qml:
```diff
@@ -10,14 +10,15 @@ import utils 1.0
     id: root
 
-    property int maximumLength: 10
+    property int maximumLength: 0
```
Does it make sense to keep this property if we will not have a "maximum amount"??

friofry replied:
I'm not sure it will be needed in the future. That's why I left it optional. I can remove this property

friofry replied:
fixed ✅

caybro commented on ui/imports/shared/controls/AmountInput.qml:
```diff
@@ -72,18 +72,20 @@ Input {
+            const fractionalPartLength = LocaleUtils.fractionalPartLength(amountNumber)
+            if (fractionalPartLength > root.tokenDecimals) {
+                d.amount = "0"
+                root.validationError = qsTr("Max. %1 decimal places for this asset").arg(root.tokenDecimals)
```
```suggestion
                root.validationError = qsTr("Max %1 decimal place(s) for this asset", "", root.tokenDecimals)
```

caybro replied:
Needs a proper plural form

friofry replied:
Done

caybro reviewed (commented):
Otherwise LGTM

kounkou commented on storybook/pages/AirdropsSettingsPanelPage.qml:
```diff
@@ -250,6 +253,19 @@ SplitView {
+
+                TextField {
+                    id: decimalsText
+                    text: "2"
```
I believe it might be better to have a constant for this? It is also repeated in L183 of storybook/pages/HoldingsDropdownPage.qml wdyt?

friofry replied:
I think it's a common approach for Storybook to hardcode test data into "<something>Page.qml".

Here is an example of its use:

<img width="1562" alt="image" src="https://github.com/status-im/status-desktop/assets/1083341/7a9339b9-c8ce-443f-bea2-4d7f0652ec2e">

kounkou commented on ui/imports/shared/controls/AmountInput.qml:
```diff
@@ -93,14 +95,14 @@ Input {
                              amountNumber, d.multiplierIndex)
 
-            if (root.validateMaximumAmount) {
+            if (root.validateMaximumAmount && root.maximumAmount && root.maximumAmount.length > 0) {
```
nit: The `&& root.maximumAmount` seems implicit to me what we are trying to achieve. Did you mean `Array.isArray(root.maximumAmount)` ? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray

caybro replied:
`maximumAmount` is a string, so this is correct imo

kounkou commented on ui/imports/shared/controls/Input.qml:
```diff
@@ -177,9 +177,10 @@ Item {
         id: validationErrorText
         visible: !!validationError
         text: validationError
+        anchors.left: inputField.left
```
Just a thought here if it's better to use Layouts instead of anchoring as anchoring are less flexible in general.

caybro replied:
Yeah but it's a really old component and unrelated to this change

kounkou commented:
Hi there, my comments are not blocking. But my need to fix conflicts here...

friofry commented:
> Hi there, my comments are not blocking. But my need to fix conflicts here...

@kounkou I've missed yours and @caybro's, sorry for not addressing these issues, it was unintentional

caybro reviewed (approved):
LGTM
